### PR TITLE
Store as self link in the links data block in the mongo record

### PR DESCRIPTION
**Pull request by HatiDev**
### JIRA link

https://companieshouse.atlassian.net/browse/LP-219

### Change description
LP-219 Store as self link in the links data block in the mongo record

### Work checklist

* [ ] Bug fix
* [x] New feature
* [ ] Breaking change
* [ ] Infrastructure change

### Pull request checklist

* [x] I have added unit tests for new code that I have added
* [x] I have added/updated functional tests where appropriate
* [x] The code follows our [coding standards](https://github.com/companieshouse/styleguides/blob/master/java.md)

__An exhaustive list of peer review checks can be read [here](https://github.com/companieshouse/styleguides/blob/master/java_review.md#developer-actions-prior-to-code-commitreview-started)__
### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java [hunk 1]
```diff
@@ -22,12 +25,19 @@ public LimitedPartnershipService(LimitedPartnershipMapper mapper, LimitedPartner
...
-    public String createLimitedPartnership(LimitedPartnershipSubmissionDto limitedPartnershipSubmissionDto, String requestId, String userId) {
+    public String createLimitedPartnership(Transaction transaction, LimitedPartnershipSubmissionDto limitedPartnershipSubmissionDto, String requestId, String userId) {
         ApiLogger.debug("Called createLimitedPartnership(...)");
 
         LimitedPartnershipSubmissionDao dao = mapper.dtoToDao(limitedPartnershipSubmissionDto);
         dao.setCreatedAt(LocalDateTime.now());
         dao.setUserId(userId);
+
+        // Create the self-link
+        String selfLink = String.format("/transactions/%s/limited-partnership-submissions/%s", transaction.getId(), dao.getId());
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java:36]: This doesn't match the format of the self link specified in the sub-task.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java [hunk 2]
```diff
@@ -42,11 +44,12 @@ public ResponseEntity<Object> createPartnership(
         var transactionId = 12321123;
         var logMap = new HashMap<String, Object>();
         logMap.put(URL_PARAM_TRANSACTION_ID, transactionId);
+        Transaction transaction = new Transaction();
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:47]: Rather than create a new transaction object at this point you need to update your branch from 'main' and use the transaction that's passed in to the method.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java (hunk 1 above)
Comment (dlloyd1-ch) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java:36]: dao.getId() will be null at this point as the mongo record has not been created. The record should be updated with the self link after it has been inserted.

Comment (HatiDev) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java]: edited to use correct format

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java (hunk 2 above)
Comment (HatiDev) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java]: after merge with main used transaction that is passed

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java (hunk 1 above)
Comment (HatiDev) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java]: revised the flow to enable update of self link after insertion

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java [hunk 3]
```diff
@@ -22,14 +25,24 @@ public LimitedPartnershipService(LimitedPartnershipMapper mapper, LimitedPartner
...
 
         LimitedPartnershipSubmissionDao dao = mapper.dtoToDao(limitedPartnershipSubmissionDto);
         dao.setCreatedAt(LocalDateTime.now());
         dao.setUserId(userId);
+
         LimitedPartnershipSubmissionDao insertedSubmission = repository.insert(dao);
-        ApiLogger.infoContext(requestId, String.format("Limited Partnership Submission created with limited-partnership submission id: %s", insertedSubmission.getId()));
+
+        // Create the self-link
+        var selfLink = String.format("/transactions/%s/limited_partnership/partnership/%s", transaction.getId(), dao.getId());
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java:38]: Sorry @HatiDev spotted this late on, but the URL on that sub-task isn't quite correct as we agreed 'limited_partnership' should be 'limited-partnership' (architect spec wasn't correct). Could you please change that?

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java [hunk 4]
```diff
@@ -10,6 +10,8 @@
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
 import uk.gov.companieshouse.api.model.transaction.Transaction;
+
+
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:14]: Minor. No need for two blank lines here.

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java (hunk 4 above)
Comment (HatiDev) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java]: line removed

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java (hunk 3 above)
Comment (HatiDev) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java]: used the correct format to use - instead of _ within (limited-partnership)

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java (hunk 1 above)
Comment (dlloyd1-ch) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java]: The link is being added to the DAO but it's not being saved to the database. Test any change locally to ensure that it is working as required.

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java (hunk 1 above)
Comment (HatiDev) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java]: if self-link is created and set in the dao before the dao is inserted into the repository, this ensures that the dao with the self-link is saved to the database.

Comment (dlloyd1-ch) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java]: The self link includes the submission ID. This is generated when the record is created in Mongo. So the record (DAO) needs to be inserted into the repository first, then updated with the self link and saved again.

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java (hunk 1 above)
Comment (HatiDev) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java]: have tested changes locally links now written to database after update with self link. see on [JIRA](https://companieshouse.atlassian.net/browse/LP-219)

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java [hunk 5]
```diff
@@ -47,6 +47,7 @@ public ResponseEntity<Object> createPartnership(
         var transactionId = transaction.getId();
         var logMap = new HashMap<String, Object>();
         logMap.put(URL_PARAM_TRANSACTION_ID, transactionId);
+
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:50]: There are a number of places in this PR where additional empty lines are being added. Please can they be removed. We should only ever add/have one blank line in order to improve readability.

Minor, but because it's occurred in a number of places, better to mention now.

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java [hunk 6]
```diff
@@ -49,9 +51,18 @@ public String createLimitedPartnership(Transaction transaction,
         LimitedPartnershipSubmissionDao dao = mapper.dtoToDao(limitedPartnershipSubmissionDto);
         dao.setCreatedAt(LocalDateTime.now());
         dao.setUserId(userId);
+
+        
+        // Create the self-link
+        var selfLink = String.format("/transactions/%s/limited-partnership/partnership/%s", transaction.getId(), dao.getId());
+        var links = new HashMap<String, String>();
+        links.put("self", selfLink);
+        dao.setLinks(links);
```
Comment (dlloyd1-ch) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java:60]: What is the purpose of this code block?  Looks redundant to me.

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java [hunk 7]
```diff
@@ -117,4 +127,10 @@ private LimitedPartnershipSubmissionDto createDto() {
 
         return submissionDto;
     }
+
+    private Transaction createTransaction() {
+        Transaction transaction = new Transaction();
+        transaction.setId(TRANSACTION_ID);
+        return transaction;
+    }
```
Comment (dlloyd1-ch) [inline src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:135]: This is a duplicate of method buildTransaction()

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java (hunk 6 above)
Comment (HatiDev) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java:60]: The creation and setting of the self-link are necessary steps to ensure that the DAO is fully populated with all required information before it is saved to the database. Therefore, this part of the code is not redundant but essential for the correct functioning of the createLimitedPartnership method.

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 7 above)
Comment (HatiDev) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:135]: now using buildTransaction which was imported after the merge with @mwestacott code

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java (hunk 6 above)
Comment (dlloyd1-ch) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java:60]: But you are not saving this dao object to the database. You are populating it with the self link and then doing nothing with it.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java (hunk 5 above)
Comment (HatiDev) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:50]: @mwestacott could you be more specific as you have just indicated the one just now and the ones you identified yesterday have been resolved.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java (hunk 6 above)
Comment (dlloyd1-ch) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java:60]: Just had another look. I can see that the dao is being inserted into the database - but as per previous comments, the self link needs the submissionId so the record needs to be updated with the self link **after** the records has been created. So my comment still stands this code appears to be redundant.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java [hunk 8]
```diff
@@ -49,9 +51,11 @@ public String createLimitedPartnership(Transaction transaction,
         LimitedPartnershipSubmissionDao dao = mapper.dtoToDao(limitedPartnershipSubmissionDto);
         dao.setCreatedAt(LocalDateTime.now());
         dao.setUserId(userId);
+
         LimitedPartnershipSubmissionDao insertedSubmission = repository.insert(dao);
 
         final String submissionUri = getSubmissionUri(transaction.getId(), insertedSubmission.getId());
+        updateLimitedPartnershipubmissionWithMetaData(insertedSubmission, submissionUri);
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java:58]: `updateLimitedPartnershipSubmissionWithSelfLink` spelling but I would also be more specific, since at the moment (and maybe until the end of time!) this method is only adding one field.

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java [hunk 9]
```diff
@@ -105,6 +110,11 @@ private Transaction buildTransaction() {
     private LimitedPartnershipSubmissionDao createDao() {
         LimitedPartnershipSubmissionDao dao = new LimitedPartnershipSubmissionDao();
         dao.setId(SUBMISSION_ID);
+        dao.setCreatedAt(LocalDateTime.now());
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:113]: I don't think you're doing anything with created date and user id so I wouldn't change your test with this data.

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java [hunk 10]
```diff
@@ -61,16 +64,18 @@ void givenDto_whenCreateLP_thenLPCreatedWithSubmissionIdAndTransactionUpdated()
...
-        Transaction transaction = buildTransaction();
-
         // when
-        String submissionId = service.createLimitedPartnership(transaction, limitedPartnershipSubmissionDto, REQUEST_ID, USER_ID);
+        String submissionId = service.createLimitedPartnership(buildTransaction(),limitedPartnershipSubmissionDto, REQUEST_ID, USER_ID);
+        Transaction transaction = buildTransaction();
 
         // then
         verify(mapper, times(1)).dtoToDao(limitedPartnershipSubmissionDto);
         verify(repository, times(1)).insert(limitedPartnershipSubmissionDao);
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:73]: As well as verifying the call to `insert` you need to verify the call now to `save`.

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java [hunk 11]
```diff
@@ -61,16 +64,18 @@ void givenDto_whenCreateLP_thenLPCreatedWithSubmissionIdAndTransactionUpdated()
...
         verify(transactionService, times(1)).updateTransaction(transactionApiCaptor.capture(), any());
         assertEquals(SUBMISSION_ID, submissionId);
+
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:76]: Three empty lines here.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java (hunk 5 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java]: Added a new comment showing where.

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 10 above)
Comment (HatiDev) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:73]: insert, save and captor verification added in test

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 11 above)
Comment (HatiDev) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java]: lines removed

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java [hunk 12]
```diff
@@ -61,15 +64,16 @@ void givenDto_whenCreateLP_thenLPCreatedWithSubmissionIdAndTransactionUpdated()
         when(mapper.dtoToDao(limitedPartnershipSubmissionDto)).thenReturn(limitedPartnershipSubmissionDao);
         when(repository.insert(limitedPartnershipSubmissionDao)).thenReturn(limitedPartnershipSubmissionDao);
 
-        Transaction transaction = buildTransaction();
-
         // when
-        String submissionId = service.createLimitedPartnership(transaction, limitedPartnershipSubmissionDto, REQUEST_ID, USER_ID);
+        String submissionId = service.createLimitedPartnership(buildTransaction(),limitedPartnershipSubmissionDto, REQUEST_ID, USER_ID);
+        Transaction transaction = buildTransaction();
```
Comment (dlloyd1-ch) [inline src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:69]: This needs to revert back to the original code where `buildTransaction()` is only invoked once.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java [hunk 13]
```diff
@@ -47,8 +47,7 @@ public ResponseEntity<Object> createPartnership(
         var transactionId = transaction.getId();
         var logMap = new HashMap<String, Object>();
         logMap.put(URL_PARAM_TRANSACTION_ID, transactionId);
-
-        try {
+    try {
```
Comment (dlloyd1-ch) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:50]: Incorrect indentation. Revert the file as this is the only change.

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java [hunk 14]
```diff
@@ -35,8 +35,8 @@ public class LimitedPartnershipServiceTest {
     private static final String USER_ID = "xbJf0l";
     private static final String SUBMISSION_ID = "abc-123";
     private static final String REQUEST_ID = "fd4gld5h3jhh";
-    private static final String TRANSACTION_ID = "12321123";
-
+    private static final String TRANSACTION_ID = "txn-456";
+
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:39]: Spaces here and an additional line added on 124. We went through checking for these kinds of things @HatiDev . I know they are minor, but if you are missing these kinds of things in GitHub after pushing, perhaps you are missing other things?

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 10 above)
Comment (mwestacott) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:73]: You need to now use the captor to validate that the link set on the submission is correct. See line 84 for an example of how that's working for the transaction captor.

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 14 above)
Comment (HatiDev) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:41]: spaces removed

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 12 above)
Comment (HatiDev) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:70]: reverted back to one invocation
```suggestion
        Transaction transaction = buildTransaction();
```

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 10 above)
Comment (HatiDev) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:75]: using captor to validate links in submission

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 14 above)
Comment (mwestacott) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:41]: There are now spaces on line 50 of the `PartnershipController` class. Did you check in GitHub after pushing?

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java [hunk 15]
```diff
@@ -47,7 +47,7 @@ public ResponseEntity<Object> createPartnership(
         var transactionId = transaction.getId();
         var logMap = new HashMap<String, Object>();
         logMap.put(URL_PARAM_TRANSACTION_ID, transactionId);
-
+
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:50]: Spaces only added here. You can just revert all the changes to this file.

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipControllerTest.java [hunk 16]
```diff
@@ -48,7 +48,6 @@ void init() {
         limitedPartnershipSubmissionDto = new LimitedPartnershipSubmissionDto();
         limitedPartnershipSubmissionDto.setData(data);
     }
-
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipControllerTest.java:51]: Why has this empty line be removed? I would just revert the changes to this file.

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java [hunk 17]
```diff
@@ -61,11 +66,10 @@ void givenDto_whenCreateLP_thenLPCreatedWithSubmissionIdAndTransactionUpdated()
         when(mapper.dtoToDao(limitedPartnershipSubmissionDto)).thenReturn(limitedPartnershipSubmissionDao);
         when(repository.insert(limitedPartnershipSubmissionDao)).thenReturn(limitedPartnershipSubmissionDao);
 
-        Transaction transaction = buildTransaction();
-
         // when
+        Transaction transaction = buildTransaction();
         String submissionId = service.createLimitedPartnership(transaction, limitedPartnershipSubmissionDto, REQUEST_ID, USER_ID);
-
+
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:72]: More spaces.

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java [hunk 18]
```diff
@@ -75,8 +79,8 @@ void givenDto_whenCreateLP_thenLPCreatedWithSubmissionIdAndTransactionUpdated()
         // assert transaction resources are updated appropriately
         Transaction sentTransaction = transactionApiCaptor.getValue();
         assertEquals(limitedPartnershipSubmissionDto.getData().getPartnershipName(), sentTransaction.getCompanyName());
-        assertNull(sentTransaction.getCompanyNumber());
-        String submissionUri = String.format("/transactions/%s/limited-partnership/%s", transaction.getId(), limitedPartnershipSubmissionDao.getId());
+        assertNull(sentTransaction.getCompanyNumber());
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:82]: Spaces at the end of the line here and on line 83.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java [hunk 19]
```diff
@@ -102,4 +105,10 @@ private boolean hasExistingLimitedPartnershipSubmission(Transaction transaction)
         }
         return false;
     }
+
+    void updateLimitedPartnershipSubmissionWithSelfLink(LimitedPartnershipSubmissionDao submission,
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java:109]: This method should remain `private` as only called from within this class. The unit-testing has to be done through the public `createLimitedPartnership()` method.

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java [hunk 20]
```diff
@@ -96,6 +100,29 @@ void givenTransactionAlreadyAssociatedWithAnLP_whenCreateLP_thenServiceException
...
+        
+        // When
+        Transaction transaction = buildTransaction();
+        LimitedPartnershipSubmissionDao submission = new LimitedPartnershipSubmissionDao();
+        String submissionUri = String.format("/transactions/%s/limited-partnership/%s", transaction.getId(), limitedPartnershipSubmissionDao.getId());
+                
+        service.updateLimitedPartnershipSubmissionWithSelfLink(submission, submissionUri);
+
+        //then
+        ArgumentCaptor<LimitedPartnershipSubmissionDao> submissionCaptor = ArgumentCaptor.forClass(LimitedPartnershipSubmissionDao.class);
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:116]: These checks look correct but they needed to be done in the method where you had them originally. The `updateLimitedPartnershipSubmissionWithSelfLink()` method should be private.

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java [hunk 21]
```diff
@@ -117,4 +144,5 @@ private LimitedPartnershipSubmissionDto createDto() {
 
         return submissionDto;
     }
+
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:147]: An extra blank line.

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 10 above)
Comment (mwestacott) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:75]: You were nearly there but now the checks have moved to a new test method.

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 10 above)
Comment (HatiDev) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:74]: captor, insert and save checks returned to old test method

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 14 above)
Comment (HatiDev) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java]: Partnership class reverted

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java (hunk 13 above)
Comment (HatiDev) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java]: file reverted

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java (hunk 15 above)
Comment (HatiDev) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java]: file reverted

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 17 above)
Comment (HatiDev) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java]: removed space

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 18 above)
Comment (HatiDev) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java]: space removed

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java (hunk 19 above)
Comment (HatiDev) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java]: method made private testing through createLimitedPartnership() method

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 20 above)
Comment (HatiDev) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java]: checks returned to original method, updateLimitedPartnershipSubmissionWithSelfLink() now private

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 21 above)
Comment (HatiDev) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java]: line removed

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java (hunk 12 above)
Comment (mwestacott) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:69]: This should move out of the `// when` block and back up a few lines, to where it was originally (it's part of the test fixture and not the method/code under test).

Comment (HatiDev) [reply src/test/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipServiceTest.java:68]: original line returned

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Review by dlloyd1-ch — APPROVED** [review]

**Review by mwestacott — APPROVED** [review]
